In [1]:
import pandas as pd
import numpy as np

# Exercise 1:

Create a `DataFrame` from a dictionary.

In [2]:
data_dict1 = {'A': np.random.randint(1, 101, size = 5), 'B': np.random.randint(1, 101, size = 5), 'C': np.random.randint(1, 101, size = 5)}

In [3]:
df1 = pd.DataFrame(data_dict1)

In [4]:
df1.head()

A   B   C
0  39  12  63
1  51  47  47
2  25  89  35
3  54  55  92
4  72  42  54

# Exercise 2:

Write a Pandas program to create and display a DataFrame from a specified dictionary data which has the index labels.

Sample Python dictionary data and list labels:

`exam_data = {'name': ['Anastasia', 'Dima', 'Katherine', 'James', 'Emily', 'Michael', 'Matthew', 'Laura', 'Kevin', 'Jonas'],
'score': [12.5, 9, 16.5, np.nan, 9, 20, 14.5, np.nan, 8, 19],
'attempts': [1, 3, 2, 3, 2, 3, 1, 1, 2, 1],
'qualify': ['yes', 'no', 'yes', 'no', 'no', 'yes', 'yes', 'no', 'no', 'yes']}
labels = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']`

Expected Output:<br>
attempts name qualify score<br>
a 1 Anastasia yes 12.5 <br>
b 3 Dima no 9.0<br>
.... i 2 Kevin no 8.0 <br>
j 1 Jonas yes 19.0<br>

In [5]:
exam_data = {
    'name': ['Anastasia', 'Dima', 'Katherine', 'James', 'Emily', 'Michael', 'Matthew', 'Laura', 'Kevin', 'Jonas'],
    'score': [12.5, 9, 16.5, np.nan, 9, 20, 14.5, np.nan, 8, 19],
    'attempts': [1, 3, 2, 3, 2, 3, 1, 1, 2, 1],
    'qualify': ['yes', 'no', 'yes', 'no', 'no', 'yes', 'yes', 'no', 'no', 'yes']
}

labels = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']

In [6]:
df2 = pd.DataFrame(data=exam_data, index=labels)

In [7]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
Index: 10 entries, a to j
Data columns (total 4 columns):
name        10 non-null object
score       8 non-null float64
attempts    10 non-null int64
qualify     10 non-null object
dtypes: float64(1), int64(1), object(2)
memory usage: 400.0+ bytes


In [8]:
df2.head()

name  score  attempts qualify
a  Anastasia   12.5         1     yes
b       Dima    9.0         3      no
c  Katherine   16.5         2     yes
d      James    NaN         3      no
e      Emily    9.0         2      no

# Exercise 3:

Write a Pandas program to display a summary of the basic information about a specified DataFrame and its data.

Sample Python dictionary data and list labels:

`exam_data = {'name': ['Anastasia', 'Dima', 'Katherine', 'James', 'Emily', 'Michael', 'Matthew', 'Laura', 'Kevin', 'Jonas'],
'score': [12.5, 9, 16.5, np.nan, 9, 20, 14.5, np.nan, 8, 19],
'attempts': [1, 3, 2, 3, 2, 3, 1, 1, 2, 1],
'qualify': ['yes', 'no', 'yes', 'no', 'no', 'yes', 'yes', 'no', 'no', 'yes']}
labels = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']`

Expected Output:<br>
Summary of the basic information about this DataFrame and its data:<br>
<class 'pandas.core.frame.DataFrame'><br>
Index: 10 entries, a to j<br>
Data columns (total 4 columns):<br>
.... dtypes: float64(1), int64(1), object(2)<br>
memory usage: 400.0+ bytes<br>
None<br>

In [9]:
df3 = df2.copy()

In [10]:
df3.info()

<class 'pandas.core.frame.DataFrame'>
Index: 10 entries, a to j
Data columns (total 4 columns):
name        10 non-null object
score       8 non-null float64
attempts    10 non-null int64
qualify     10 non-null object
dtypes: float64(1), int64(1), object(2)
memory usage: 400.0+ bytes


# Exercise 4:

Write a Pandas program to get the first 3 rows of a given DataFrame.

Sample Python dictionary data and list labels:

`exam_data = {'name': ['Anastasia', 'Dima', 'Katherine', 'James', 'Emily', 'Michael', 'Matthew', 'Laura', 'Kevin', 'Jonas'],
'score': [12.5, 9, 16.5, np.nan, 9, 20, 14.5, np.nan, 8, 19],
'attempts': [1, 3, 2, 3, 2, 3, 1, 1, 2, 1],
'qualify': ['yes', 'no', 'yes', 'no', 'no', 'yes', 'yes', 'no', 'no', 'yes']}
labels = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']`

Expected Output:<br>
First three rows of the data frame:<br>
attempts name qualify score <br>
a 1 Anastasia yes 12.5<br>
b 3 Dima no 9.0 <br>
c 2 Katherine yes 16.5<br>

In [11]:
df4 = df3.copy()

In [12]:
df4.head(3)

name  score  attempts qualify
a  Anastasia   12.5         1     yes
b       Dima    9.0         3      no
c  Katherine   16.5         2     yes

In [13]:
df4.iloc[:3]

name  score  attempts qualify
a  Anastasia   12.5         1     yes
b       Dima    9.0         3      no
c  Katherine   16.5         2     yes

# Exercise 5:

Write a Pandas program to select the 'name' and 'score' columns from the following DataFrame.

Sample Python dictionary data and list labels:

`exam_data = {'name': ['Anastasia', 'Dima', 'Katherine', 'James', 'Emily', 'Michael', 'Matthew', 'Laura', 'Kevin', 'Jonas'],
'score': [12.5, 9, 16.5, np.nan, 9, 20, 14.5, np.nan, 8, 19],
'attempts': [1, 3, 2, 3, 2, 3, 1, 1, 2, 1],
'qualify': ['yes', 'no', 'yes', 'no', 'no', 'yes', 'yes', 'no', 'no', 'yes']}
labels = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']`

Expected Output:<br>
Select specific columns:<br>
name score <br>
a Anastasia 12.5<br> 
b Dima 9.0 <br>
c Katherine 16.5<br>
... h Laura NaN <br>
i Kevin 8.0 <br>
j Jonas 19.0<br>

In [14]:
df5 = df4.copy()

In [15]:
df5.loc[:,['name', 'score']]

name  score
a  Anastasia   12.5
b       Dima    9.0
c  Katherine   16.5
d      James    NaN
e      Emily    9.0
f    Michael   20.0
g    Matthew   14.5
h      Laura    NaN
i      Kevin    8.0
j      Jonas   19.0

# Exercise 6:

Write a Pandas program to select the specified columns and rows from a given data frame.

Sample Python dictionary data and list labels:
Select 'name' and 'score' columns in rows 1, 3, 5, 6 from the following data frame.

`exam_data = {'name': ['Anastasia', 'Dima', 'Katherine', 'James', 'Emily', 'Michael', 'Matthew', 'Laura', 'Kevin', 'Jonas'],
'score': [12.5, 9, 16.5, np.nan, 9, 20, 14.5, np.nan, 8, 19],
'attempts': [1, 3, 2, 3, 2, 3, 1, 1, 2, 1],
'qualify': ['yes', 'no', 'yes', 'no', 'no', 'yes', 'yes', 'no', 'no', 'yes']}
labels = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']`

Expected Output:<br>
Select specific columns and rows:<br>
name score <br>
b Dima 9.0 <br>
d James NaN <br>
f Michael 20.0 <br>
g Matthew 14.5 <br>

In [16]:
df6 = df5.copy()

In [17]:
df6.loc[['b', 'd', 'f', 'g'], ['name', 'score']]

name  score
b     Dima    9.0
d    James    NaN
f  Michael   20.0
g  Matthew   14.5

# Exercise 7:

Write a Pandas program to select the rows where the number of attempts in the examination is greater than 2.

Sample Python dictionary data and list labels:

`exam_data = {'name': ['Anastasia', 'Dima', 'Katherine', 'James', 'Emily', 'Michael', 'Matthew', 'Laura', 'Kevin', 'Jonas'],
'score': [12.5, 9, 16.5, np.nan, 9, 20, 14.5, np.nan, 8, 19],
'attempts': [1, 3, 2, 3, 2, 3, 1, 1, 2, 1],
'qualify': ['yes', 'no', 'yes', 'no', 'no', 'yes', 'yes', 'no', 'no', 'yes']}
labels = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']`

Expected Output:<br>
Select specific columns and rows:<br>
name score <br>
b Dima 9.0 <br>
d James NaN <br>
f Michael 20.0<br>

In [18]:
df7 = df6.copy()

In [19]:
df7[df7.score > 2][['name', 'score']]

name  score
a  Anastasia   12.5
b       Dima    9.0
c  Katherine   16.5
e      Emily    9.0
f    Michael   20.0
g    Matthew   14.5
i      Kevin    8.0
j      Jonas   19.0

# Exercise 8:

Write a Pandas program to count the number of rows and columns of a DataFrame.

Sample Python dictionary data and list labels:

`exam_data = {'name': ['Anastasia', 'Dima', 'Katherine', 'James', 'Emily', 'Michael', 'Matthew', 'Laura', 'Kevin', 'Jonas'],
'score': [12.5, 9, 16.5, np.nan, 9, 20, 14.5, np.nan, 8, 19],
'attempts': [1, 3, 2, 3, 2, 3, 1, 1, 2, 1],
'qualify': ['yes', 'no', 'yes', 'no', 'no', 'yes', 'yes', 'no', 'no', 'yes']}
labels = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']`

Expected Output:<br>
Number of Rows: 10 <br>
Number of Columns: 4<br>

In [20]:
df8 = df7.copy()

In [21]:
'Number of rows: {}'.format(df8.shape[0])

'Number of rows: 10'

In [22]:
'Number of columns: {}'.format(df8.shape[1])

'Number of columns: 4'

# Exercise 9:

Write a Pandas program to select the rows where the score is missing, i.e. is NaN.

Sample Python dictionary data and list labels:

`exam_data = {'name': ['Anastasia', 'Dima', 'Katherine', 'James', 'Emily', 'Michael', 'Matthew', 'Laura', 'Kevin', 'Jonas'],
'score': [12.5, 9, 16.5, np.nan, 9, 20, 14.5, np.nan, 8, 19],
'attempts': [1, 3, 2, 3, 2, 3, 1, 1, 2, 1],
'qualify': ['yes', 'no', 'yes', 'no', 'no', 'yes', 'yes', 'no', 'no', 'yes']}
labels = ['a', 'b', 'c', 'd', 'e', 'f', 'g', labels = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']`

Expected Output:<br>
Rows where score is missing:<br>
attempts name qualify score<br>
d 3 James no NaN<br>
h 1 Laura no NaN<br>

In [23]:
df9 = df8.copy()

In [24]:
df9[df9['score'].isnull()]

name  score  attempts qualify
d  James    NaN         3      no
h  Laura    NaN         1      no

# Exercise 10:

Write a Pandas program to select the rows the score is between 15 and 20 (inclusive).

Sample Python dictionary data and list labels:

`exam_data = {'name': ['Anastasia', 'Dima', 'Katherine', 'James', 'Emily', 'Michael', 'Matthew', 'Laura', 'Kevin', 'Jonas'],
'score': [12.5, 9, 16.5, np.nan, 9, 20, 14.5, np.nan, 8, 19],
'attempts': [1, 3, 2, 3, 2, 3, 1, 1, 2, 1],
'qualify': ['yes', 'no', 'yes', 'no', 'no', 'yes', 'yes', 'no', 'no', 'yes']}
labels = ['a', 'b', 'c', 'd', 'e', 'f', 'g', labels = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']`

Expected Output:<br>
Rows where score between 15 and 20 (inclusive):<br>
attempts name qualify score <br>
c 2 Katherine yes 16.5 <br>
f 3 Michael yes 20.0 <br>
j 1 Jonas yes 19.0<br>

In [25]:
df10 = df9.copy()

In [26]:
df10[(df10['score'] >= 15) & (df10['score'] <= 20)]

name  score  attempts qualify
c  Katherine   16.5         2     yes
f    Michael   20.0         3     yes
j      Jonas   19.0         1     yes

# Exercise 11:

Write a Pandas program to select the rows where number of attempts in the examination is less than 2 and score greater than 15.

Sample Python dictionary data and list labels:

`exam_data = {'name': ['Anastasia', 'Dima', 'Katherine', 'James', 'Emily', 'Michael', 'Matthew', 'Laura', 'Kevin', 'Jonas'],
'score': [12.5, 9, 16.5, np.nan, 9, 20, 14.5, np.nan, 8, 19],
'attempts': [1, 3, 2, 3, 2, 3, 1, 1, 2, 1],
'qualify': ['yes', 'no', 'yes', 'no', 'no', 'yes', 'yes', 'no', 'no', 'yes']}
labels = ['a', 'b', 'c', 'd', 'e', 'f', 'g', labels = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']`

Expected Output:<br>
Rows where score between 15 and 20 (inclusive):<br>
attempts name qualify score <br>
c 2 Katherine yes 16.5 <br>
j 1 Jonas yes 19.0<br>

In [27]:
df11 = df10.copy()

In [28]:
df11[(df11.attempts < 2) & (df11.score > 15)]

name  score  attempts qualify
j  Jonas   19.0         1     yes

# Exercise 12:

Write a Pandas program to change the score in row 'd' to 11.5. 

Sample Python dictionary data and list labels:

`exam_data = {'name': ['Anastasia', 'Dima', 'Katherine', 'James', 'Emily', 'Michael', 'Matthew', 'Laura', 'Kevin', 'Jonas'],
'score': [12.5, 9, 16.5, np.nan, 9, 20, 14.5, np.nan, 8, 19],
'attempts': [1, 3, 2, 3, 2, 3, 1, 1, 2, 1],
'qualify': ['yes', 'no', 'yes', 'no', 'no', 'yes', 'yes', 'no', 'no', 'yes']}
labels = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']`

Expected Output:<br>
Change the score in row 'd' to 11.5: <br>
attempts name qualify score<br>
a 1 Anastasia yes 12.5<br>
b 3 Dima no 9.0<br>
c 2 Katherine yes 16.5<br>
... <br>
i 2 Kevin no 8.0 <br>
j 1 Jonas yes 19.0<br>

In [29]:
df12 = df11.copy()

In [30]:
df12.at['d', 'score'] = 11.5

In [31]:
df12

name  score  attempts qualify
a  Anastasia   12.5         1     yes
b       Dima    9.0         3      no
c  Katherine   16.5         2     yes
d      James   11.5         3      no
e      Emily    9.0         2      no
f    Michael   20.0         3     yes
g    Matthew   14.5         1     yes
h      Laura    NaN         1      no
i      Kevin    8.0         2      no
j      Jonas   19.0         1     yes

# Exercise 13:

Write a Pandas program to calculate the sum of the examination attempts by the students.

Sample Python dictionary data and list labels:

`exam_data = {'name': ['Anastasia', 'Dima', 'Katherine', 'James', 'Emily', 'Michael', 'Matthew', 'Laura', 'Kevin', 'Jonas'],
'score': [12.5, 9, 16.5, np.nan, 9, 20, 14.5, np.nan, 8, 19],
'attempts': [1, 3, 2, 3, 2, 3, 1, 1, 2, 1],
'qualify': ['yes', 'no', 'yes', 'no', 'no', 'yes', 'yes', 'no', 'no', 'yes']}
labels = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']`

Expected Output:<br>
Sum of the examination attempts by the students:<br>
19<br>

In [32]:
df13 = df12.copy()

In [33]:
'Sum of the examination attempts by the students: {}'.format(df13.attempts.sum())

'Sum of the examination attempts by the students: 19'

# Exercise 14:

Write a Pandas program to calculate the mean score for each different student in DataFrame.

Sample Python dictionary data and list labels:

`exam_data = {'name': ['Anastasia', 'Dima', 'Katherine', 'James', 'Emily', 'Michael', 'Matthew', 'Laura', 'Kevin', 'Jonas'],
'score': [12.5, 9, 16.5, np.nan, 9, 20, 14.5, np.nan, 8, 19],
'attempts': [1, 3, 2, 3, 2, 3, 1, 1, 2, 1],
'qualify': ['yes', 'no', 'yes', 'no', 'no', 'yes', 'yes', 'no', 'no', 'yes']}
labels = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']`

Expected Output:<br>
Mean score for each different student in data frame: <br>
13.5625<br>

In [34]:
df14 = df2.copy()

In [35]:
df14.score.mean()

13.5625

# Exercise 15:

Write a Pandas program to append a new row 'k' to data frame with given values for each column. Now delete the new row and return the original DataFrame.

Sample Python dictionary data and list labels:

`exam_data = {'name': ['Anastasia', 'Dima', 'Katherine', 'James', 'Emily', 'Michael', 'Matthew', 'Laura', 'Kevin', 'Jonas'],
'score': [12.5, 9, 16.5, np.nan, 9, 20, 14.5, np.nan, 8, 19],
'attempts': [1, 3, 2, 3, 2, 3, 1, 1, 2, 1],
'qualify': ['yes', 'no', 'yes', 'no', 'no', 'yes', 'yes', 'no', 'no', 'yes']}
labels = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']`

Values for each column will be:<br>
name : "Suresh", score: 15.5, attempts: 1, qualify: "yes", label: "k"<br>
Expected Output:<br>
Append a new row:<br>
Print all records after insert a new record:<br>
attempts name qualify score<br>
a 1 Anastasia yes 12.5<br>
b 3 Dima no 9.0<br>
......<br>
j 1 Jonas yes 19.0<br>
k 1 Suresh yes 15.5<br>
Delete the new row and display the original rows:<br>
attempts name qualify score<br>
a 1 Anastasia yes 12.5<br>
b 3 Dima no 9.0<br>
........<br>
i 2 Kevin no 8.0<br>
j 1 Jonas yes 19.0<br>

In [36]:
df15 = df14.copy()

In [37]:
new_student = {'name': 'Julio Ordonez', 'score': 13.9, 'attempts': 1, 'qualify': 'yes'}

In [38]:
df15.loc['k'] = pd.Series(new_student)

In [39]:
df15

name  score  attempts qualify
a      Anastasia   12.5         1     yes
b           Dima    9.0         3      no
c      Katherine   16.5         2     yes
d          James    NaN         3      no
e          Emily    9.0         2      no
f        Michael   20.0         3     yes
g        Matthew   14.5         1     yes
h          Laura    NaN         1      no
i          Kevin    8.0         2      no
j          Jonas   19.0         1     yes
k  Julio Ordonez   13.9         1     yes

In [40]:
df15.drop(['k'])

name  score  attempts qualify
a  Anastasia   12.5         1     yes
b       Dima    9.0         3      no
c  Katherine   16.5         2     yes
d      James    NaN         3      no
e      Emily    9.0         2      no
f    Michael   20.0         3     yes
g    Matthew   14.5         1     yes
h      Laura    NaN         1      no
i      Kevin    8.0         2      no
j      Jonas   19.0         1     yes

# Exercise 16:

Write a Pandas program to sort the DataFrame first by 'name' in descending order, then by 'score' in ascending order.

Sample Python dictionary data and list labels:

`exam_data = {'name': ['Anastasia', 'Dima', 'Katherine', 'James', 'Emily', 'Michael', 'Matthew', 'Laura', 'Kevin', 'Jonas'],
'score': [12.5, 9, 16.5, np.nan, 9, 20, 14.5, np.nan, 8, 19],
'attempts': [1, 3, 2, 3, 2, 3, 1, 1, 2, 1],
'qualify': ['yes', 'no', 'yes', 'no', 'no', 'yes', 'yes', 'no', 'no', 'yes']}
labels = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']`

Values for each column will be:<br>
name : "Suresh", score: 15.5, attempts: 1, qualify: "yes", label: "k"<br>
Expected Output:<br>
Sort the data frame first by 'name' in descending order, then by 'score' in ascending order: <br>
attempts name qualify score <br>
a 1 Anastasia yes 12.5<br>
b 3 Dima no 9.0<br>
..... <br>
i 2 Kevin no 8.0 <br>
j 1 Jonas yes 19.0<br>

In [41]:
df16 = df15.copy()

In [42]:
df16.sort_values(by=['name', 'score'], ascending=[False, True])

name  score  attempts qualify
f        Michael   20.0         3     yes
g        Matthew   14.5         1     yes
h          Laura    NaN         1      no
i          Kevin    8.0         2      no
c      Katherine   16.5         2     yes
k  Julio Ordonez   13.9         1     yes
j          Jonas   19.0         1     yes
d          James    NaN         3      no
e          Emily    9.0         2      no
b           Dima    9.0         3      no
a      Anastasia   12.5         1     yes

# Exercise 17:

Write a Pandas program to replace the 'qualify' column contains the values 'yes' and 'no' with True and False.

Sample Python dictionary data and list labels:

`exam_data = {'name': ['Anastasia', 'Dima', 'Katherine', 'James', 'Emily', 'Michael', 'Matthew', 'Laura', 'Kevin', 'Jonas'],
'score': [12.5, 9, 16.5, np.nan, 9, 20, 14.5, np.nan, 8, 19],
'attempts': [1, 3, 2, 3, 2, 3, 1, 1, 2, 1],
'qualify': ['yes', 'no', 'yes', 'no', 'no', 'yes', 'yes', 'no', 'no', 'yes']}
labels = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']`

Expected Output:<br>
Replace the 'qualify' column contains the values 'yes' and 'no' with T<br>
rue and False: <br>
attempts name qualify score <br>
a 1 Anastasia True 12.5 <br>
b 3 Dima False 9.0 <br>
...... <br>
i 2 Kevin False 8.0 <br>
j 1 Jonas True 19.0<br>

In [43]:
df17 = df16.copy()

In [44]:
df17['qualify'] = df17['qualify'].replace(to_replace='yes', value=True).replace(to_replace='no', value=False)

In [45]:
df17.head()

name  score  attempts  qualify
a  Anastasia   12.5         1     True
b       Dima    9.0         3    False
c  Katherine   16.5         2     True
d      James    NaN         3    False
e      Emily    9.0         2    False

In [46]:
df17.qualify.dtype

dtype('bool')

In [47]:
df17.info()

<class 'pandas.core.frame.DataFrame'>
Index: 11 entries, a to k
Data columns (total 4 columns):
name        11 non-null object
score       9 non-null float64
attempts    11 non-null int64
qualify     11 non-null bool
dtypes: bool(1), float64(1), int64(1), object(1)
memory usage: 363.0+ bytes


# Exercise 18:

Write a Pandas program to change the name 'James' to 'Suresh' in name column of the DataFrame. 

Sample Python dictionary data and list labels:

`exam_data = {'name': ['Anastasia', 'Dima', 'Katherine', 'James', 'Emily', 'Michael', 'Matthew', 'Laura', 'Kevin', 'Jonas'],
'score': [12.5, 9, 16.5, np.nan, 9, 20, 14.5, np.nan, 8, 19],
'attempts': [1, 3, 2, 3, 2, 3, 1, 1, 2, 1],
'qualify': ['yes', 'no', 'yes', 'no', 'no', 'yes', 'yes', 'no', 'no', 'yes']}
labels = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']`

Expected Output:<br>
Change the name 'James' to \?Suresh\?:<br>
attempts name qualify score<br>
a 1 Anastasia yes 12.5 <br>
b 3 Dima no 9.0<br>
....... <br>
i 2 Kevin no 8.0<br>
j 1 Jonas yes 19.0<br>

In [48]:
df18 = df16.copy()

In [49]:
df18['name'] = df18['name'].replace(to_replace='James', value='Suresh')

In [50]:
df18['name']

a        Anastasia
b             Dima
c        Katherine
d           Suresh
e            Emily
f          Michael
g          Matthew
h            Laura
i            Kevin
j            Jonas
k    Julio Ordonez
Name: name, dtype: object

# Exercise 19

Write a Pandas program to delete the 'attempts' column from the DataFrame.

Sample Python dictionary data and list labels:

`exam_data = {'name': ['Anastasia', 'Dima', 'Katherine', 'James', 'Emily', 'Michael', 'Matthew', 'Laura', 'Kevin', 'Jonas'],
'score': [12.5, 9, 16.5, np.nan, 9, 20, 14.5, np.nan, 8, 19],
'attempts': [1, 3, 2, 3, 2, 3, 1, 1, 2, 1],
'qualify': ['yes', 'no', 'yes', 'no', 'no', 'yes', 'yes', 'no', 'no', 'yes']}
labels = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']`

Expected Output:<br>
Delete the 'attempts' column from the data frame: <br>
name qualify score<br>
a Anastasia yes 12.5 <br>
b Dima no 9.0 <br>
..... <br>
i Kevin no 8.0<br>
j Jonas yes 19.0<br>

In [51]:
df19 = df16.copy()

In [52]:
df19.drop(labels=['attempts'], axis=1)

name  score qualify
a      Anastasia   12.5     yes
b           Dima    9.0      no
c      Katherine   16.5     yes
d          James    NaN      no
e          Emily    9.0      no
f        Michael   20.0     yes
g        Matthew   14.5     yes
h          Laura    NaN      no
i          Kevin    8.0      no
j          Jonas   19.0     yes
k  Julio Ordonez   13.9     yes

In [53]:
# df.pop('attempts')

# Exercise 20:

Write a Pandas program to insert a new column in existing DataFrame.

Sample Python dictionary data and list labels:

`exam_data = {'name': ['Anastasia', 'Dima', 'Katherine', 'James', 'Emily', 'Michael', 'Matthew', 'Laura', 'Kevin', 'Jonas'],
'score': [12.5, 9, 16.5, np.nan, 9, 20, 14.5, np.nan, 8, 19],
'attempts': [1, 3, 2, 3, 2, 3, 1, 1, 2, 1],
'qualify': ['yes', 'no', 'yes', 'no', 'no', 'yes', 'yes', 'no', 'no', 'yes']}
labels = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']`

Expected Output:<br>
New DataFrame after inserting the 'color' column <br>
attempts name qualify score color<br>
a 1 Anastasia yes 12.5 Red<br>
b 3 Dima no 9.0 Blue<br>
....... <br>
i 2 Kevin no 8.0 Green <br>
j 1 Jonas yes 19.0 Red<br>

In [54]:
colors = ['blue', 'red', 'green', 'yellow', 'pink', 'magenta', 'black', 'white', 'purple', 'gray', 'orange']

In [55]:
df20 = df16.copy()

In [56]:
df20['color'] = pd.Series(colors, index=df20.index)

In [57]:
df20.head(7)

name  score  attempts qualify    color
a  Anastasia   12.5         1     yes     blue
b       Dima    9.0         3      no      red
c  Katherine   16.5         2     yes    green
d      James    NaN         3      no   yellow
e      Emily    9.0         2      no     pink
f    Michael   20.0         3     yes  magenta
g    Matthew   14.5         1     yes    black

# Exercise 21:

Write a Pandas program to iterate over rows in a DataFrame.

Sample Python dictionary data and list labels:

`exam_data = [{'name':'Anastasia', 'score':12.5}, {'name':'Dima','score':9}, {'name':'Katherine','score':16.5}]`

Expected Output:<br>
Anastasia 12.5<br>
Dima 9.0 <br>
Katherine 16.5<br>

In [58]:
exam_data = [{'name':'Anastasia', 'score':12.5}, {'name':'Dima','score':9}, {'name':'Katherine','score':16.5}]

In [59]:
df21 = pd.DataFrame(data=exam_data)

In [60]:
df21.head()

name  score
0  Anastasia   12.5
1       Dima    9.0
2  Katherine   16.5

In [61]:
for index, row in df21.iterrows():
    print(row['name'], row['score'])

Anastasia 12.5
Dima 9.0
Katherine 16.5


# Exercise 22:

Write a Pandas program to get list from DataFrame column headers.

Sample Python dictionary data and list labels:

`exam_data = {'name': ['Anastasia', 'Dima', 'Katherine', 'James', 'Emily', 'Michael', 'Matthew', 'Laura', 'Kevin', 'Jonas'],
'score': [12.5, 9, 16.5, np.nan, 9, 20, 14.5, np.nan, 8, 19],
'attempts': [1, 3, 2, 3, 2, 3, 1, 1, 2, 1],
'qualify': ['yes', 'no', 'yes', 'no', 'no', 'yes', 'yes', 'no', 'no', 'yes']}
labels = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']`

Expected Output:

['attempts', 'name', 'qualify', 'score']

In [62]:
df22 = df6.copy()

In [63]:
list(df22.columns.values)

['name', 'score', 'attempts', 'qualify']

# Exercise 23:

Write a Pandas program to rename columns of a given DataFrame

Sample data:

Original DataFrame<br>
col1 col2 col3<br>
0 1 4 7<br>
1 2 5 8<br>
2 3 6 9<br>

New DataFrame after renaming columns:<br>
Column1 Column2 Column3<br>
0 1 4 7<br>
1 2 5 8<br>
2 3 6 9<br>

In [64]:
data = {'col1': [1, 2, 3], 'col2': [4, 5, 6], 'col3': [7, 8, 9]}

In [65]:
df23 = pd.DataFrame(data=data)

In [66]:
df23

col1  col2  col3
0     1     4     7
1     2     5     8
2     3     6     9

In [67]:
columns_rename = {'col1': 'Column1', 'col2': 'Column2', 'col3': 'Column3'}

In [68]:
df23_copy = df23.copy()

In [69]:
df23.rename(columns=columns_rename)

Column1  Column2  Column3
0        1        4        7
1        2        5        8
2        3        6        9

# Exercise 24:

Write a Pandas program to select rows from a given DataFrame based on values in some columns. 

Sample data:

Original DataFrame<br>
col1 col2 col3<br>
0 1 4 7<br>
1 4 5 8<br>
2 3 6 9<br>
3 4 7 0<br>
4 5 8 1<br>

Rows for colum1 value == 4<br>
col1 col2 col3<br>
1 4 5 8<br>
3 4 7 0<br>

In [70]:
data = {'col1': [1, 4, 3, 4, 5], 'col2': [4, 5, 6, 7, 8], 'col3': [7, 8, 9 , 0 ,1]}

In [71]:
df24 = pd.DataFrame(data=data)

In [72]:
df24.head()

col1  col2  col3
0     1     4     7
1     4     5     8
2     3     6     9
3     4     7     0
4     5     8     1

In [73]:
df24.loc[df24.col1 == 4]

col1  col2  col3
1     4     5     8
3     4     7     0

# Exercise 25:

Write a Pandas program to change the order of a DataFrame columns.

Sample data:

Original DataFrame<br>
col1 col2 col3<br>
0 1 4 7<br>
1 4 5 8<br>
2 3 6 9<br>
3 4 7 0<br>
4 5 8 1<br>

After altering col1 and col3<br>
col3 col2 col1<br>
0 7 4 1<br>
1 8 5 4<br>
2 9 6 3<br>
3 0 7 4<br>
4 1 8 5<br>

In [74]:
df25 = df24.copy()

In [75]:
columns_new_order = sorted(df25.columns.tolist(), reverse=True)

In [76]:
columns_new_order

['col3', 'col2', 'col1']

In [77]:
df25 = df25[columns_new_order]

In [78]:
df25

col3  col2  col1
0     7     4     1
1     8     5     4
2     9     6     3
3     0     7     4
4     1     8     5

# Exercise 26:

Write a Pandas program to add one row in an existing DataFrame.

Sample data:

Original DataFrame<br>
col1 col2 col3<br>
0 1 4 7<br>
1 4 5 8<br>
2 3 6 9<br>
3 4 7 0<br>
4 5 8 1<br>

After add one row:<br>
col1 col2 col3<br>
0 1 4 7<br>
1 4 5 8<br>
2 3 6 9<br>
3 4 7 0<br>
4 5 8 1<br>
5 10 11 12<br>

In [79]:
df26 = df24.copy()

In [80]:
df26.loc[6] = [10, 11, 12]

In [81]:
df26

col1  col2  col3
0     1     4     7
1     4     5     8
2     3     6     9
3     4     7     0
4     5     8     1
6    10    11    12

In [82]:
# df = {'col1': 10, 'col2': 11, 'col3': 12}
# df26 = df26.append(df, ignore_index=True)

# Exercise 27:

Write a Pandas program to write a DataFrame to CSV file using tab separator.

Sample data:

Original DataFrame<br>
col1 col2 col3<br>
0 1 4 7<br>
1 4 5 8<br>
2 3 6 9<br>
3 4 7 0<br>
4 5 8 1<br>

Data from new_file.csv file:

col1\tcol2\tcol3<br>
0 1\t4\t7<br>
1 4\t5\t8<br>
2 3\t6\t9<br>
3 4\t7\t0<br>
4 5\t8\t1<br>

In [83]:
df27 = df24.copy()

In [84]:
df27.to_csv('df25_new_file.csv', sep='\t', index=False)

# Exercise 28:

Write a Pandas program to count city wise number of people from a given of data set (city, name of the person).

Sample data:

city Number of people<br>
0 California 4<br>
1 Georgia 2<br>
2 Los Angeles 4<br>

In [85]:
df28 = pd.DataFrame({'name': ['Anastasia', 'Dima', 'Katherine', 'James', 'Emily', 'Michael', 'Matthew', 'Laura', 'Kevin', 'Jonas'],
'city': ['California', 'Los Angeles', 'California', 'California', 'California', 'Los Angeles', 'Los Angeles', 'Georgia', 'Georgia', 'Los Angeles']})

In [86]:
df28

name         city
0  Anastasia   California
1       Dima  Los Angeles
2  Katherine   California
3      James   California
4      Emily   California
5    Michael  Los Angeles
6    Matthew  Los Angeles
7      Laura      Georgia
8      Kevin      Georgia
9      Jonas  Los Angeles

In [87]:
df28.groupby('city').size().reset_index(name='Number of people')

city  Number of people
0   California                 4
1      Georgia                 2
2  Los Angeles                 4

# Exercise 29:

Write a Pandas program to delete DataFrame row(s) based on given column value.

Sample data:

Original DataFrame<br>
col1 col2 col3<br>
0 1 4 7<br>
1 4 5 8<br>
2 3 6 9<br>
3 4 7 0<br>
4 5 8 1<br>

New DataFrame<br>
col1 col2 col3<br>
0 1 4 7<br>
2 3 6 9<br>
3 4 7 0<br>
4 5 8 1<br>

In [88]:
df29 = df24.copy()

In [89]:
df29

col1  col2  col3
0     1     4     7
1     4     5     8
2     3     6     9
3     4     7     0
4     5     8     1

In [90]:
df29.drop(df29.loc[df29.col1 == 4].index.tolist())

col1  col2  col3
0     1     4     7
2     3     6     9
4     5     8     1

# Exercise 30:

Write a Pandas program to widen output display to see more columns.

Sample data:

Original DataFrame<br>
col1 col2 col3<br>
0 1 4 7<br>
1 4 5 8<br>
2 3 6 9<br>
3 4 7 0<br>
4 5 8 1<br>

In [91]:
d = {'col1': [1, 4, 3, 4, 5], 'col2': [4, 5, 6, 7, 8], 'col3': [7, 8, 9, 0, 1]}
df = pd.DataFrame(data=d)

pd.set_option('display.max_rows', 500)
pd.set_option('display.max_columns', 500)
pd.set_option('display.width', 1000)

df

col1  col2  col3
0     1     4     7
1     4     5     8
2     3     6     9
3     4     7     0
4     5     8     1

# Exercise 31:

Write a Pandas program to select a row of series/dataframe by given integer index.

Sample data:

Original DataFrame<br>
col1 col2 col3<br>
0 1 4 7<br>
1 4 5 8<br>
2 3 6 9<br>
3 4 7 0<br>
4 5 8 1<br>

Index-2: Details<br>
col1 col2 col3<br>
2 3 6 9<br>

In [92]:
df31 = df24.copy()

In [93]:
df31

col1  col2  col3
0     1     4     7
1     4     5     8
2     3     6     9
3     4     7     0
4     5     8     1

In [94]:
df31.iloc[[2]]

col1  col2  col3
2     3     6     9

# Exercise 32:

Write a Pandas program to replace all the NaN values with Zero's in a column of a dataframe.

Sample data:

Original DataFrame<br>
attempts name qualify score<br>
0 1 Anastasia yes 12.5<br>
1 3 Dima no 9.0<br>
2 2 Katherine yes 16.5<br>
3 3 James no NaN<br>
4 2 Emily no 9.0<br>
5 3 Michael yes 20.0<br>
6 1 Matthew yes 14.5<br>
7 1 Laura no NaN<br>
8 2 Kevin no 8.0<br>
9 1 Jonas yes 19.0<br>

New DataFrame replacing all NaN with 0:

attempts name qualify score<br>
0 1 Anastasia yes 12.5<br>
1 3 Dima no 9.0<br>
2 2 Katherine yes 16.5<br>
3 3 James no 0.0<br>
4 2 Emily no 9.0<br>
5 3 Michael yes 20.0<br>
6 1 Matthew yes 14.5<br>
7 1 Laura no 0.0<br>
8 2 Kevin no 8.0<br>
9 1 Jonas yes 19.0<br>

In [95]:
df32 = df2.copy()
df32

name  score  attempts qualify
a  Anastasia   12.5         1     yes
b       Dima    9.0         3      no
c  Katherine   16.5         2     yes
d      James    NaN         3      no
e      Emily    9.0         2      no
f    Michael   20.0         3     yes
g    Matthew   14.5         1     yes
h      Laura    NaN         1      no
i      Kevin    8.0         2      no
j      Jonas   19.0         1     yes

In [96]:
df32.score.fillna(0, inplace=True)

In [97]:
df32

name  score  attempts qualify
a  Anastasia   12.5         1     yes
b       Dima    9.0         3      no
c  Katherine   16.5         2     yes
d      James    0.0         3      no
e      Emily    9.0         2      no
f    Michael   20.0         3     yes
g    Matthew   14.5         1     yes
h      Laura    0.0         1      no
i      Kevin    8.0         2      no
j      Jonas   19.0         1     yes

# Exercise 33:

Write a Pandas program to convert index in a column of the given dataframe.

Sample data:

Original DataFrame<br>
attempts name qualify score<br>
0 1 Anastasia yes 12.5<br>
1 3 Dima no 9.0<br>
2 2 Katherine yes 16.5<br>
3 3 James no NaN<br>
4 2 Emily no 9.0<br>
5 3 Michael yes 20.0<br>
6 1 Matthew yes 14.5<br>
7 1 Laura no NaN<br>
8 2 Kevin no 8.0<br>
9 1 Jonas yes 19.0<br>

After converting index in a column:

index attempts name qualify score<br>
0 0 1 Anastasia yes 12.5<br>
1 1 3 Dima no 9.0<br>
2 2 2 Katherine yes 16.5<br>
3 3 3 James no NaN<br>
4 4 2 Emily no 9.0<br>
5 5 3 Michael yes 20.0<br>
6 6 1 Matthew yes 14.5<br>
7 7 1 Laura no NaN<br>
8 8 2 Kevin no 8.0<br>
9 9 1 Jonas yes 19.0<br>

Hiding index:

index attempts name qualify score<br>
0 1 Anastasia yes 12.5<br>
1 3 Dima no 9.0<br>
2 2 Katherine yes 16.5<br>
3 3 James no NaN<br>
4 2 Emily no 9.0<br>
5 3 Michael yes 20.0<br>
6 1 Matthew yes 14.5<br>
7 1 Laura no NaN<br>
8 2 Kevin no 8.0<br>
9 1 Jonas yes 19.0<br>

In [98]:
df33 = df2.copy()
df33

name  score  attempts qualify
a  Anastasia   12.5         1     yes
b       Dima    9.0         3      no
c  Katherine   16.5         2     yes
d      James    NaN         3      no
e      Emily    9.0         2      no
f    Michael   20.0         3     yes
g    Matthew   14.5         1     yes
h      Laura    NaN         1      no
i      Kevin    8.0         2      no
j      Jonas   19.0         1     yes

In [99]:
df33.index = range(0, 10)
df33

name  score  attempts qualify
0  Anastasia   12.5         1     yes
1       Dima    9.0         3      no
2  Katherine   16.5         2     yes
3      James    NaN         3      no
4      Emily    9.0         2      no
5    Michael   20.0         3     yes
6    Matthew   14.5         1     yes
7      Laura    NaN         1      no
8      Kevin    8.0         2      no
9      Jonas   19.0         1     yes

In [100]:
df33.reset_index(level=0, inplace=True)
df33

index       name  score  attempts qualify
0      0  Anastasia   12.5         1     yes
1      1       Dima    9.0         3      no
2      2  Katherine   16.5         2     yes
3      3      James    NaN         3      no
4      4      Emily    9.0         2      no
5      5    Michael   20.0         3     yes
6      6    Matthew   14.5         1     yes
7      7      Laura    NaN         1      no
8      8      Kevin    8.0         2      no
9      9      Jonas   19.0         1     yes

In [101]:
print(df33.to_string(index=False))

index       name  score  attempts qualify
    0  Anastasia   12.5         1     yes
    1       Dima    9.0         3      no
    2  Katherine   16.5         2     yes
    3      James    NaN         3      no
    4      Emily    9.0         2      no
    5    Michael   20.0         3     yes
    6    Matthew   14.5         1     yes
    7      Laura    NaN         1      no
    8      Kevin    8.0         2      no
    9      Jonas   19.0         1     yes


# Exercise 34:

Write a Pandas program to set a given value for particular cell in  DataFrame using index value.

Sample data:

Original DataFrame<br>
attempts name qualify score<br>
0 1 Anastasia yes 12.5<br>
1 3 Dima no 9.0<br>
2 2 Katherine yes 16.5<br>
3 3 James no NaN<br>
4 2 Emily no 9.0<br>
5 3 Michael yes 20.0<br>
6 1 Matthew yes 14.5<br>
7 1 Laura no NaN<br>
8 2 Kevin no 8.0<br>
9 1 Jonas yes 19.0<br>

Set a given value for particular cell in the DataFrame

attempts name qualify score<br>
0 1 Anastasia yes 12.5<br>
1 3 Dima no 9.0<br>
2 2 Katherine yes 16.5<br>
3 3 James no NaN<br>
4 2 Emily no 9.0<br>
5 3 Michael yes 20.0<br>
6 1 Matthew yes 14.5<br>
7 1 Laura no NaN<br>
8 2 Kevin no 10.2<br>
9 1 Jonas yes 19.0<br>

In [102]:
df34 = df2.copy()
df34

name  score  attempts qualify
a  Anastasia   12.5         1     yes
b       Dima    9.0         3      no
c  Katherine   16.5         2     yes
d      James    NaN         3      no
e      Emily    9.0         2      no
f    Michael   20.0         3     yes
g    Matthew   14.5         1     yes
h      Laura    NaN         1      no
i      Kevin    8.0         2      no
j      Jonas   19.0         1     yes

In [103]:
print(df34.loc['a', 'score'])
df34.loc['a', 'score'] = 10.5
print(df34.loc['a', 'score'])

12.5
10.5


In [104]:
# df34.set_value('a', 'score', 10.5)

# Exercise 35:

Write a Pandas program to count the NaN values in one or more columns in DataFrame.

Sample data:

Original DataFrame<br>
attempts name qualify score<br>
0 1 Anastasia yes 12.5<br>
1 3 Dima no 9.0<br>
2 2 Katherine yes 16.5<br>
3 3 James no NaN<br>
4 2 Emily no 9.0<br>
5 3 Michael yes 20.0<br>
6 1 Matthew yes 14.5<br>
7 1 Laura no NaN<br>
8 2 Kevin no 8.0<br>
9 1 Jonas yes 19.0<br>
Number of NaN values in one or more columns:<br>
2

In [105]:
exam_data = {
    'name': ['Anastasia', 'Dima', 'Katherine', 'James', 'Emily', 'Michael', 'Matthew', 'Laura', 'Kevin', 'Jonas'],
    'score': [12.5, 9, 16.5, np.nan, 9, 20, 14.5, np.nan, 8, 19],
    'attempts': [1, 3, 2, 3, 2, 3, 1, 1, 2, 1],
    'qualify': ['yes', 'no', 'yes', 'no', 'no', 'yes', 'yes', 'no', 'no', 'yes']
}

In [106]:
df35 = pd.DataFrame(data=exam_data)

In [107]:
df35

name  score  attempts qualify
0  Anastasia   12.5         1     yes
1       Dima    9.0         3      no
2  Katherine   16.5         2     yes
3      James    NaN         3      no
4      Emily    9.0         2      no
5    Michael   20.0         3     yes
6    Matthew   14.5         1     yes
7      Laura    NaN         1      no
8      Kevin    8.0         2      no
9      Jonas   19.0         1     yes

In [108]:
df35.isnull().sum().max()

2

In [109]:
# df35.isnull().values.sum()

# Exercise 36:

Write a Pandas program to drop a list of rows from a specified DataFrame.

Sample data:

Original DataFrame<br>
col1 col2 col3<br>
0 1 4 7<br>
1 4 5 8<br>
2 3 6 9<br>
3 4 7 0<br>
4 5 8 1<br>

New DataFrame after removing 2nd & 4th rows:

col1 col2 col3<br>
0 1 4 7<br>
1 4 5 8<br>
3 4 7 0<br>

In [110]:
df36 = df24.copy()
df36

col1  col2  col3
0     1     4     7
1     4     5     8
2     3     6     9
3     4     7     0
4     5     8     1

In [111]:
df36.drop([2, 4])

col1  col2  col3
0     1     4     7
1     4     5     8
3     4     7     0

# Exercise 37:

Write a Pandas program to reset index in a given DataFrame.

Sample data:

Original DataFrame<br>
attempts name qualify score<br>
0 1 Anastasia yes 12.5<br>
1 3 Dima no 9.0<br>
2 2 Katherine yes 16.5<br>
3 3 James no NaN<br>
4 2 Emily no 9.0<br>
5 3 Michael yes 20.0<br>
6 1 Matthew yes 14.5<br>
7 1 Laura no NaN<br>
8 2 Kevin no 8.0<br>
9 1 Jonas yes 19.0<br>

After removing first and second rows

attempts name qualify score<br>
2 2 Katherine yes 16.5<br>
3 3 James no NaN<br>
4 2 Emily no 9.0<br>
5 3 Michael yes 20.0<br>
6 1 Matthew yes 14.5<br>
7 1 Laura no NaN<br>
8 2 Kevin no 8.0<br>
9 1 Jonas yes 19.0<br>

Reset the Index:

index attempts name qualify score<br>
0 2 2 Katherine yes 16.5<br>
1 3 3 James no NaN<br>
2 4 2 Emily no 9.0<br>
3 5 3 Michael yes 20.0<br>
4 6 1 Matthew yes 14.5<br>
5 7 1 Laura no NaN<br>
6 8 2 Kevin no 8.0<br>
7 9 1 Jonas yes 19.0<br>

In [112]:
df37 = df35.copy()
df37

name  score  attempts qualify
0  Anastasia   12.5         1     yes
1       Dima    9.0         3      no
2  Katherine   16.5         2     yes
3      James    NaN         3      no
4      Emily    9.0         2      no
5    Michael   20.0         3     yes
6    Matthew   14.5         1     yes
7      Laura    NaN         1      no
8      Kevin    8.0         2      no
9      Jonas   19.0         1     yes

In [113]:
df37.drop(df37.index[[0, 1]], inplace=True)
df37

name  score  attempts qualify
2  Katherine   16.5         2     yes
3      James    NaN         3      no
4      Emily    9.0         2      no
5    Michael   20.0         3     yes
6    Matthew   14.5         1     yes
7      Laura    NaN         1      no
8      Kevin    8.0         2      no
9      Jonas   19.0         1     yes

In [114]:
df37.reset_index(inplace=True)

In [115]:
df37

index       name  score  attempts qualify
0      2  Katherine   16.5         2     yes
1      3      James    NaN         3      no
2      4      Emily    9.0         2      no
3      5    Michael   20.0         3     yes
4      6    Matthew   14.5         1     yes
5      7      Laura    NaN         1      no
6      8      Kevin    8.0         2      no
7      9      Jonas   19.0         1     yes

# Exercise 38:

Write a Pandas program to devide a DataFrame in a given ratio.

Sample data:

Original DataFrame:<br>
0 1<br>
0 0.316147 -0.767359<br>
1 -0.813410 -2.522672<br>
2 0.869615 1.194704<br>
3 -0.892915 -0.055133<br>
4 -0.341126 0.518266<br>
5 1.857342 1.361229<br>
6 -0.044353 -1.205002<br>
7 -0.726346 -0.535147<br>
8 -1.350726 0.563117<br>
9 1.051666 -0.441533<br>

70% of the said DataFrame:

0 1<br>
8 -1.350726 0.563117<br>
2 0.869615 1.194704<br>
5 1.857342 1.361229<br>
6 -0.044353 -1.205002<br>
3 -0.892915 -0.055133<br>
1 -0.813410 -2.522672<br>
0 0.316147 -0.767359<br>

30% of the said DataFrame:

0 1<br>
4 -0.341126 0.518266<br>
7 -0.726346 -0.535147<br>
9 1.051666 -0.441533<br>

In [116]:
df38 = pd.DataFrame(np.random.randn(10, 2))
df38

0         1
0 -1.511951  0.625402
1 -0.584012  0.056919
2 -1.544986  0.452018
3  1.036425 -0.440734
4  1.491535 -0.143255
5  2.229280  0.339932
6 -0.180800 -0.231024
7  0.230025 -0.276672
8  0.412612 -1.111939
9 -0.703635 -1.295270

In [117]:
df38_70percent = df38.sample(frac=0.7, random_state=10)
df38_70percent

0         1
8  0.412612 -1.111939
2 -1.544986  0.452018
5  2.229280  0.339932
6 -0.180800 -0.231024
3  1.036425 -0.440734
1 -0.584012  0.056919
0 -1.511951  0.625402

In [118]:
df38_30percent = df38.drop(df38_70percent.index)
df38_30percent

0         1
4  1.491535 -0.143255
7  0.230025 -0.276672
9 -0.703635 -1.295270

# Exercise 39:

Write a Pandas program to combining two series into a DataFrame.

Sample data:

Data Series:<br>
0 100<br>
1 200<br>
2 python<br>
3 300.12<br>
4 400<br>
dtype: object

0 10<br>
1 20<br>
2 php<br>
3 30.12<br>
4 40<br>
dtype: object

New DataFrame combining two series:

0 1<br>
0 100 10<br>
1 200 20<br>
2 python php<br>
3 300.12 30.12<br>
4 400 40<br>

In [119]:
data1 = [100, 200, 'python', 300.12, 400]
data2 = [10, 20, 'php', 30.12, 40]

series1 = pd.Series(data=data1)
series2 = pd.Series(data=data2)

In [120]:
series1

0       100
1       200
2    python
3    300.12
4       400
dtype: object

In [121]:
series2

0       10
1       20
2      php
3    30.12
4       40
dtype: object

In [122]:
df39 = pd.concat([series1, series2], axis=1)
df39

0      1
0     100     10
1     200     20
2  python    php
3  300.12  30.12
4     400     40

# Exercise 40:

Write a Pandas program to shuffle a given DataFrame rows.

Sample data:

Original DataFrame:

attempts name qualify score<br>
0 1 Anastasia yes 12.5<br>
1 3 Dima no 9.0<br>
2 2 Katherine yes 16.5<br>
3 3 James no NaN<br>
4 2 Emily no 9.0<br>
5 3 Michael yes 20.0<br>
6 1 Matthew yes 14.5<br>
7 1 Laura no NaN<br>
8 2 Kevin no 8.0<br>
9 1 Jonas yes 19.0<br>

New DataFrame:

attempts name qualify score<br>
5 3 Michael yes 20.0<br>
0 1 Anastasia yes 12.5<br>
9 1 Jonas yes 19.0<br>
6 1 Matthew yes 14.5<br>
7 1 Laura no NaN<br>
1 3 Dima no 9.0<br>
3 3 James no NaN<br>
4 2 Emily no 9.0<br>
8 2 Kevin no 8.0<br>
2 2 Katherine yes 16.5<br>

In [123]:
df40 = df35.copy()
df40

name  score  attempts qualify
0  Anastasia   12.5         1     yes
1       Dima    9.0         3      no
2  Katherine   16.5         2     yes
3      James    NaN         3      no
4      Emily    9.0         2      no
5    Michael   20.0         3     yes
6    Matthew   14.5         1     yes
7      Laura    NaN         1      no
8      Kevin    8.0         2      no
9      Jonas   19.0         1     yes

In [124]:
indexes = df40.index.values.tolist()

In [125]:
df40 = df40.reindex(np.random.permutation(df40.index))
df40

name  score  attempts qualify
4      Emily    9.0         2      no
8      Kevin    8.0         2      no
6    Matthew   14.5         1     yes
0  Anastasia   12.5         1     yes
7      Laura    NaN         1      no
3      James    NaN         3      no
1       Dima    9.0         3      no
9      Jonas   19.0         1     yes
5    Michael   20.0         3     yes
2  Katherine   16.5         2     yes

In [126]:
# df40 = df40.sample(frac=1)

# Exercise 41:

Write a Pandas program to convert DataFrame column type from string to datetime.

Sample data:

String Date:

0 3/11/2000<br>
1 3/12/2000<br>
2 3/13/2000<br>
dtype: object<br>

Original DataFrame (string to datetime):

0<br>
0 2000-03-11<br>
1 2000-03-12<br>
2 2000-03-13<br>

In [127]:
data = ['3/11/2000', '3/12/2000', '3/13/2000']

In [128]:
dates = pd.Series(data=data)

In [129]:
dates

0    3/11/2000
1    3/12/2000
2    3/13/2000
dtype: object

In [130]:
df41 = pd.DataFrame(data=dates)

In [131]:
df41

0
0  3/11/2000
1  3/12/2000
2  3/13/2000

In [132]:
df41[0] = pd.to_datetime(df41[0], errors='coerce')
df41

0
0 2000-03-11
1 2000-03-12
2 2000-03-13

In [133]:
df41.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3 entries, 0 to 2
Data columns (total 1 columns):
0    3 non-null datetime64[ns]
dtypes: datetime64[ns](1)
memory usage: 104.0 bytes


# Exercise 42:

Write a Pandas program to rename a specific column name in a given DataFrame.

Sample data:

Original DataFrame

col1 col2 col3<br>
0 1 4 7<br>
1 2 5 8<br>
2 3 6 9<br>

New DataFrame after renaming second column:

col1 Column2 col3<br>
0 1 4 7<br>
1 2 5 8<br>
2 3 6 9<br>

In [134]:
data = np.arange(1, 10, dtype=np.int).reshape((3, 3)).T

In [135]:
columns = ['col1', 'col2', 'col3']

In [136]:
df42 = pd.DataFrame(data=data, columns=columns)
df42

col1  col2  col3
0     1     4     7
1     2     5     8
2     3     6     9

In [137]:
df42 = df42.rename({'col2': 'Column2'}, axis=1)

In [138]:
df42

col1  Column2  col3
0     1        4     7
1     2        5     8
2     3        6     9

In [139]:
# df42 = df42.rename(columns={'col2': 'Column2'})

# Exercise 43:

Write a Pandas program to get a list of a specified column of a DataFrame.

Sample data:<br>

Powered by 

Original DataFrame

col1 col2 col3<br>
0 1 4 7<br>
1 2 5 8<br>
2 3 6 9<br>

Col2 of the DataFrame to list:

[4, 5, 6]

In [140]:
data = np.arange(1, 10, dtype=np.int).reshape((3, 3)).T
columns = ['col{}'.format(col) for col in range(1, 4)]

In [141]:
df43 = pd.DataFrame(data=data, columns=columns)
df43

col1  col2  col3
0     1     4     7
1     2     5     8
2     3     6     9

In [142]:
data_col2 = df43.col2.values.tolist()
data_col2

[4, 5, 6]

# Exercise 44:

Write a Pandas program to create a DataFrame from a Numpy array and specify the index column and column headers.

Sample Output:

Column1 Column2 Column3<br>
Index1 0 0.0 0.0<br>
Index2 0 0.0 0.0<br>
Index3 0 0.0 0.0<br>
.........<br>
Index12 0 0.0 0.0<br>
Index13 0 0.0 0.0<br>
Index14 0 0.0 0.0<br>
Index15 0 0.0 0.0<br>

In [143]:
indexes = ['Index{}'.format(i) for i in range(1, 16)]
columns = ['Column{}'.format(j) for j in range(1, 4)]

data_Column1 = np.zeros(15, dtype=np.int)
data_Column2 = np.zeros(15, dtype=np.float)
data_Column3 = np.zeros(15, dtype=np.float)

series1 = pd.Series(data_Column1, index=indexes, name=columns[0])
series2 = pd.Series(data_Column2, index=indexes, name=columns[1])
series3 = pd.Series(data_Column3, index=indexes, name=columns[2])

In [144]:
df44 = pd.concat([series1, series2, series3], axis=1)
df44

Column1  Column2  Column3
Index1         0      0.0      0.0
Index2         0      0.0      0.0
Index3         0      0.0      0.0
Index4         0      0.0      0.0
Index5         0      0.0      0.0
Index6         0      0.0      0.0
Index7         0      0.0      0.0
Index8         0      0.0      0.0
Index9         0      0.0      0.0
Index10        0      0.0      0.0
Index11        0      0.0      0.0
Index12        0      0.0      0.0
Index13        0      0.0      0.0
Index14        0      0.0      0.0
Index15        0      0.0      0.0

In [145]:
# Alternative solution:

In [146]:
dtype = [('Column1','int32'), ('Column2','float32'), ('Column3','float32')]
values = np.zeros(15, dtype=dtype)
index = ['Index'+str(i) for i in range(1, len(values)+1)]
df44_alt = pd.DataFrame(values, index=index)
df44_alt

Column1  Column2  Column3
Index1         0      0.0      0.0
Index2         0      0.0      0.0
Index3         0      0.0      0.0
Index4         0      0.0      0.0
Index5         0      0.0      0.0
Index6         0      0.0      0.0
Index7         0      0.0      0.0
Index8         0      0.0      0.0
Index9         0      0.0      0.0
Index10        0      0.0      0.0
Index11        0      0.0      0.0
Index12        0      0.0      0.0
Index13        0      0.0      0.0
Index14        0      0.0      0.0
Index15        0      0.0      0.0

# Exercise 45:

Write a Pandas program to find the row for where the value of a given column is maximum.

Sample Output:

Original DataFrame

col1 col2 col3<br>
0 1 4 7<br>
1 2 5 8<br>
2 3 6 12<br>
3 4 9 1<br>
4 7 5 11<br>

Row where col1 has maximum value:<br>
4

Row where col2 has maximum value:<br>
3

Row where col3 has maximum value:<br>
2

In [147]:
columns = ['col{}'.format(j) for j in range(1, 4)]

series1 = pd.Series(data=[1, 2, 3, 4, 7])
series2 = pd.Series(data=[4, 5, 6, 9, 5])
series3 = pd.Series(data=[7, 8, 12, 1, 11])

df45 = pd.DataFrame(columns=columns)

df45['col1'] = series1
df45['col2'] = series2
df45['col3'] = series3

df45

col1  col2  col3
0     1     4     7
1     2     5     8
2     3     6    12
3     4     9     1
4     7     5    11

In [148]:
for col in df45:
    print('Row where {} has maximum value: {}'.format(col, df45.index[df45[col] == df45[col].max()][0]))

Row where col1 has maximum value: 4
Row where col2 has maximum value: 3
Row where col3 has maximum value: 2


In [149]:
# Alternative solution:

In [150]:
for col in df45:
    print('Row where {} has maximum value: {}'.format(col, df45[col].idxmax()))

Row where col1 has maximum value: 4
Row where col2 has maximum value: 3
Row where col3 has maximum value: 2


# Exercise 46:

Write a Pandas program to check whether a given column is present in a DataFrame or not

Sample data:

Original DataFrame

col1 col2 col3<br>
0 1 4 7<br>
1 2 5 8<br>
2 3 6 12<br>
3 4 9 1<br>
4 7 5 11<br>

Col4 is not present in DataFrame.

Col1 is present in DataFrame.

In [151]:
df46 = df45.copy()

column_name = 'col4'

'{} is {} present in DataFrame'.format(column_name, '' if column_name in df46 else 'not')

'col4 is not present in DataFrame'

In [152]:
column_name = 'col1'

'{} is {} present in DataFrame'.format(column_name, '' if column_name in df46 else 'not')

'col1 is  present in DataFrame'

# Exercise 47:

Write a Pandas program to get the specified row value of a given DataFrame.

Sample data:

Original DataFrame

col1 col2 col3<br>
0 1 4 7<br>
1 2 5 8<br>
2 3 6 12<br>
3 4 9 1<br>
4 7 5 11<br>

Value of Row<br>
col1 1<br>
col2 4<br>
col3 7<br>
Name: 0, dtype: int64<br>

Value of Row4<br>
col1 4<br>
col2 9<br>
col3 1<br>
Name: 3, dtype: int64<br>

In [153]:
df47 = df45.copy()
df47

col1  col2  col3
0     1     4     7
1     2     5     8
2     3     6    12
3     4     9     1
4     7     5    11

In [154]:
df.iloc[0]

col1    1
col2    4
col3    7
Name: 0, dtype: int64

In [155]:
df.iloc[3]

col1    4
col2    7
col3    0
Name: 3, dtype: int64

# Exercise 48:

Write a Pandas program to get the datatypes of columns of a DataFrame.

Sample data:
Original DataFrame:

attempts name qualify score<br>
0 1 Anastasia yes 12.5<br>
1 3 Dima no 9.0<br>
.......<br>
8 2 Kevin no 8.0<br>
9 1 Jonas yes 19.0

Data types of the columns of the said DataFrame:

attempts int64<br>
name object<br>
qualify object<br>
score float64<br>
dtype: object

In [156]:
df48 = df35.copy()
df48

name  score  attempts qualify
0  Anastasia   12.5         1     yes
1       Dima    9.0         3      no
2  Katherine   16.5         2     yes
3      James    NaN         3      no
4      Emily    9.0         2      no
5    Michael   20.0         3     yes
6    Matthew   14.5         1     yes
7      Laura    NaN         1      no
8      Kevin    8.0         2      no
9      Jonas   19.0         1     yes

In [157]:
df48.dtypes

name         object
score       float64
attempts      int64
qualify      object
dtype: object

# Exercise 49:

Write a Pandas program to append data to an empty DataFrame.

Sample data:

Original DataFrame:

After appending some data:

col1 col2<br>
0 0 0 <br>
1 1 1 <br>
2 2 2

In [158]:
columns = ['col1', 'col2']

df49 = pd.DataFrame(columns=columns)
df49

Empty DataFrame
Columns: [col1, col2]
Index: []

In [159]:
df49.loc[len(df49)] = [0, 0]
df49.loc[len(df49)] = [1, 1]
df49.loc[len(df49)] = [2, 2]

df49

col1 col2
0    0    0
1    1    1
2    2    2

# Exercise 50:

Write a Pandas program to sort a given DataFrame by two or more columns.

Sample data:

Original DataFrame:

attempts name qualify score<br>
0 1 Anastasia yes 12.5 <br>
1 3 Dima no 9.0<br>
........<br>
8 2 Kevin no 8.0<br>
9 1 Jonas yes 19.0<br>

Sort the above DataFrame on attempts, name:

attempts name qualify score<br>
0 1 Anastasia yes 12.5<br>
9 1 Jonas yes 19.0<br>
7 1 Laura no NaN<br>
6 1 Matthew yes 14.5<br>
4 2 Emily no 9.0<br>
2 2 Katherine yes 16.5<br>
8 2 Kevin no 8.0<br>
1 3 Dima no 9.0<br>
3 3 James no NaN<br>
5 3 Michael yes 20.0

In [160]:
df50 = df35.copy()
df50

name  score  attempts qualify
0  Anastasia   12.5         1     yes
1       Dima    9.0         3      no
2  Katherine   16.5         2     yes
3      James    NaN         3      no
4      Emily    9.0         2      no
5    Michael   20.0         3     yes
6    Matthew   14.5         1     yes
7      Laura    NaN         1      no
8      Kevin    8.0         2      no
9      Jonas   19.0         1     yes

In [161]:
df50.sort_values(by=['attempts', 'name'])

name  score  attempts qualify
0  Anastasia   12.5         1     yes
9      Jonas   19.0         1     yes
7      Laura    NaN         1      no
6    Matthew   14.5         1     yes
4      Emily    9.0         2      no
2  Katherine   16.5         2     yes
8      Kevin    8.0         2      no
1       Dima    9.0         3      no
3      James    NaN         3      no
5    Michael   20.0         3     yes

# Exercise 51

Write a Pandas program to convert the datatype of a given column (floats to ints).

Sample data:

Original DataFrame:

attempts name qualify score<br>
0 1 Anastasia yes 12.50<br>
1 3 Dima no 9.10<br>
......<br>
8 2 Kevin no 8.80<br>
9 1 Jonas yes 19.13<br>

Data types of the columns of the said DataFrame:

attempts int64<br>
name object<br>
qualify object<br>
score float64<br>
dtype: object<br>

Now change the Data type of 'score' column from float to int:

attempts name qualify score<br>
0 1 Anastasia yes 12<br>
1 3 Dima no 9<br>
2 2 Katherine yes 16<br>
3 3 James no 12<br>
4 2 Emily no 9<br>
5 3 Michael yes 20<br>
6 1 Matthew yes 14<br>
7 1 Laura no 11<br>
8 2 Kevin no 8<br>
9 1 Jonas yes 19<br>

Data types of the columns of the DataFrame now:

attempts int64<br>
name object<br>
qualify object<br>
score int64<br>
dtype: object

In [162]:
df51 = df35.copy()
df51

name  score  attempts qualify
0  Anastasia   12.5         1     yes
1       Dima    9.0         3      no
2  Katherine   16.5         2     yes
3      James    NaN         3      no
4      Emily    9.0         2      no
5    Michael   20.0         3     yes
6    Matthew   14.5         1     yes
7      Laura    NaN         1      no
8      Kevin    8.0         2      no
9      Jonas   19.0         1     yes

In [163]:
df51.dtypes

name         object
score       float64
attempts      int64
qualify      object
dtype: object

In [164]:
df51.score = df51.score.fillna(0).astype(int)
df51.dtypes

name        object
score        int64
attempts     int64
qualify     object
dtype: object

# Exercise 52:

Write a Pandas program to remove infinite values from a given DataFrame.

Sample data:

Original DataFrame:

0<br>
0 1000.000000<br>
1 2000.000000<br>
2 3000.000000<br>
3 -4000.000000<br>
4 inf<br>
5 -inf<br>

Removing infinite values:

0<br>
0 1000.0<br>
1 2000.0<br>
2 3000.0<br>
3 -4000.0<br>
4 NaN<br>
5 NaN<br>

In [165]:
data = [1000.000000, 2000.0000000, 3000.000000, -4000.0000000, np.inf, -np.inf]

df52 = pd.DataFrame(data=data)
df52

0
0  1000.000000
1  2000.000000
2  3000.000000
3 -4000.000000
4          inf
5         -inf

In [166]:
df52 = df52.replace([np.inf, -np.inf], np.nan)
df52

0
0  1000.0
1  2000.0
2  3000.0
3 -4000.0
4     NaN
5     NaN

# Exercise 53:

Write a Pandas program to insert a given column at a specific column index in a DataFrame.

Sample data:

Original DataFrame

col2 col3<br>
0 4 7<br>
1 5 8<br>
2 6 12<br>
3 9 1<br>
4 5 11<br>

New DataFrame

col1 col2 col3<br>
0 1 4 7<br>
1 2 5 8<br>
2 3 6 12<br>
3 4 9 1<br>
4 7 5 11<br>

In [167]:
data_col2 = [4, 5, 6, 9, 5]
data_col3 = [7, 8, 12, 1, 11]

columns = ['col2', 'col3']

df53 = pd.DataFrame(columns=columns)

df53.col2 = data_col2
df53.col3 = data_col3

df53

col2  col3
0     4     7
1     5     8
2     6    12
3     9     1
4     5    11

In [168]:
index = 0
data_col1 = [1, 2, 3, 4, 7]
column_name = 'col1'

df53.insert(loc=index, column=column_name, value=data_col1)

In [169]:
df53

col1  col2  col3
0     1     4     7
1     2     5     8
2     3     6    12
3     4     9     1
4     7     5    11

# Exercise 54:

Write a Pandas program to convert a given list of lists into a Dataframe.

Sample data:

Original list of lists:

[[2, 4], [1, 3]]

New DataFrame

col1 col2<br>
0 2 4<br>
1 1 3<br>

In [170]:
matrix = [[2, 4], [1, 3]]

pd54 = pd.DataFrame(data=matrix, columns=['col1', 'col2'])

pd54

col1  col2
0     2     4
1     1     3

# Exercise 55:

Write a Pandas program to group by the first column and get second column as lists in rows.

Sample data:<br>

Original DataFrame

col1 col2<br>
0 C1 1<br>
1 C1 2<br>
2 C2 3<br>
3 C2 3<br>
4 C2 4<br>
5 C3 6<br>
6 C2 5<br>

Group on the col1:

col1<br>
C1 [1, 2]<br>
C2 [3, 3, 4, 5]<br>
C3 [6]<br>
Name: col2, dtype: object

In [171]:
data_col1 = ['C1', 'C1', 'C2', 'C2', 'C2', 'C3', 'C2']
data_col2 = [1, 2, 3, 3, 4, 6, 5]

df55 = pd.DataFrame(columns=['col1', 'col2'])

In [172]:
df55.col1 = data_col1
df55.col2 = data_col2

df55

col1  col2
0   C1     1
1   C1     2
2   C2     3
3   C2     3
4   C2     4
5   C3     6
6   C2     5

In [173]:
df55.groupby(by='col1').col2.apply(list)

col1
C1          [1, 2]
C2    [3, 3, 4, 5]
C3             [6]
Name: col2, dtype: object

# Exercise 56:

Write a Pandas program to get column index from column name of a given DataFrame.

Sample Output:

Original DataFrame

col1 col2 col3<br>
0 1 4 7<br>
1 2 5 8<br>
2 3 6 12<br>
3 4 9 1<br>
4 7 5 11<br>

Index of 'col2'
1

In [174]:
df56 = df53.copy()

df56

col1  col2  col3
0     1     4     7
1     2     5     8
2     3     6    12
3     4     9     1
4     7     5    11

In [175]:
"Index of 'col1' {}".format(df56.columns.get_loc('col2'))

"Index of 'col1' 1"

# Exercise 57:

Write a Pandas program to count number of columns of a DataFrame.

Sample Output:

Original DataFrame

col1 col2 col3<br>
0 1 4 7<br>
1 2 5 8<br>
2 3 6 12<br>
3 4 9 1<br>
4 7 5 11<br>

Number of columns: 3

In [176]:
df57 = df53.copy()

In [177]:
'Number of columns: {}'.format(df57.columns.size)

'Number of columns: 3'

# Exercise 58:

Write a Pandas program to select all columns, except one given column in a DataFrame.

Sample Output:

Original DataFrame

col1 col2 col3<br>
0 1 4 7<br>
1 2 5 8<br>
2 3 6 12<br>
3 4 9 1<br>
4 7 5 11<br>

All columns except 'col3':

col1 col2<br>
0 1 4<br>
1 2 5<br>
2 3 6<br>
3 4 9<br>
4 7 5

In [178]:
df58 = df53.copy()
df58

col1  col2  col3
0     1     4     7
1     2     5     8
2     3     6    12
3     4     9     1
4     7     5    11

In [179]:
df58[['col1', 'col2']]

col1  col2
0     1     4
1     2     5
2     3     6
3     4     9
4     7     5

In [180]:
df58.loc[:, df58.columns != 'col3']

col1  col2
0     1     4
1     2     5
2     3     6
3     4     9
4     7     5

# Exercise 59:

Write a Pandas program to get first n records of a DataFrame.

Sample Output:

Original DataFrame

col1 col2 col3<br><br>
0 1 4 7<br>
1 2 5 5<br>
2 3 6 8<br>
3 4 9 12<br>
4 7 5 1<br>
5 11 0 11<br>

First 3 rows of the said DataFrame':

col1 col2 col3<br>
0 1 4 7<br>
1 2 5 5<br>
2 3 6 8

In [181]:
df59 = df53.copy()

df59

col1  col2  col3
0     1     4     7
1     2     5     8
2     3     6    12
3     4     9     1
4     7     5    11

In [182]:
df59.loc[:2,]

col1  col2  col3
0     1     4     7
1     2     5     8
2     3     6    12

# Exercise 60:

Write a Pandas program to get last n records of a DataFrame.

Sample Output:

Original DataFrame

col1 col2 col3<br>
0 1 4 7<br>
1 2 5 5<br>
2 3 6 8<br>
3 4 9 12<br>
4 7 5 1<br>
5 11 0 11

Last 3 rows of the said DataFrame':

col1 col2 col3<br>
3 4 9 12<br>
4 7 5 1<br>
5 11 0 11

In [183]:
columns = ['col{}'.format(j) for j in range(1, 4)]

data_col1 = [1, 2, 3, 4, 7, 11]
data_col2 = [4, 5, 6, 9, 5, 0]
data_col3 = [7, 5, 8, 12, 1, 11]

df60 = pd.DataFrame(columns=columns)

df60.col1 = data_col1
df60.col2 = data_col2
df60.col3 = data_col3

df60

col1  col2  col3
0     1     4     7
1     2     5     5
2     3     6     8
3     4     9    12
4     7     5     1
5    11     0    11

In [184]:
df60.tail(3)

col1  col2  col3
3     4     9    12
4     7     5     1
5    11     0    11

# Exercise 61:

Write a Pandas program to get topmost n records within each group of a DataFrame.

Sample Output:

Original DataFrame

col1 col2 col3<br>
0 1 4 7<br>
1 2 5 5<br>
2 3 6 8<br>
3 4 9 12<br>
4 7 5 1<br>
5 11 0 11<br>

topmost n records within each group of a DataFrame:

col1 col2 col3<br>
5 11 0 11<br>
4 7 5 1<br>
3 4 9 12<br>
col1 col2 col3<br>
3 4 9 12<br>
2 3 6 8<br>
1 2 5 5<br>
4 7 5 1<br>
col1 col2 col3<br>
3 4 9 12<br>
5 11 0 11<br>
2 3 6 8

In [185]:
df61 = df60.copy()

df61

col1  col2  col3
0     1     4     7
1     2     5     5
2     3     6     8
3     4     9    12
4     7     5     1
5    11     0    11

In [186]:
df61.nlargest(3, 'col1')

col1  col2  col3
5    11     0    11
4     7     5     1
3     4     9    12

In [187]:
df61.nlargest(3, 'col2')

col1  col2  col3
3     4     9    12
2     3     6     8
1     2     5     5

In [188]:
df61.nlargest(3, 'col3')

col1  col2  col3
3     4     9    12
5    11     0    11
2     3     6     8

# Exercise 62:

Write a Pandas program to remove first n rows of a given DataFrame.

Sample Output:

Original DataFrame

col1 col2 col3<br>
0 1 4 7<br>
1 2 5 5<br>
2 3 6 8<br>
3 4 9 12<br>
4 7 5 1<br>
5 11 0 11<br>

After removing first 3 rows of the said DataFrame:

col1 col2 col3<br>
3 4 9 12<br>
4 7 5 1<br>
5 11 0 11

In [189]:
df62 = df60.copy()
df62

col1  col2  col3
0     1     4     7
1     2     5     5
2     3     6     8
3     4     9    12
4     7     5     1
5    11     0    11

In [190]:
df62.drop(range(3))

col1  col2  col3
3     4     9    12
4     7     5     1
5    11     0    11

In [191]:
# Alternate solution:

# df = df.iloc[3:]

# Exercise 63:

Write a Pandas program to remove last n rows of a given DataFrame.

Sample Output:

Original DataFrame

col1 col2 col3<br>
0 1 4 7<br>
1 2 5 5<br>
2 3 6 8<br>
3 4 9 12<br>
4 7 5 1<br>
5 11 0 11

After removing last 3 rows of the said DataFrame:

col1 col2 col3<br>
0 1 4 7<br>
1 2 5 5<br>
2 3 6 8

In [192]:
df63 = df60.copy()
df63.iloc[:-3]

col1  col2  col3
0     1     4     7
1     2     5     5
2     3     6     8